src/tokenserver/auth/browserid.rs

```diff
@@ -0,0 +1,469 @@
+            })
+            .send()
+            .map_err(|e| {
+                if e.is_connect() {
+                    // If we are unable to reach the FxA server, report a 503 to the client
+                    TokenserverError::resource_unavailable()
+                } else {
+                    // If any other error occurs during the request, report a 401 to the client
+                    TokenserverError::invalid_credentials("Unauthorized")
+                }
+            })?;
+
+        // If FxA responds with an HTTP status other than 200, report a 503 to the client
+        if response.status() != StatusCode::OK {
+            return Err(TokenserverError::resource_unavailable());
+        }
+
+        // If FxA responds with an invalid response body, report a 503 to the client
+        let response_body = response
+            .json::<VerifyResponse>()
+            .map_err(|_| TokenserverError::resource_unavailable())?;
+
+        match response_body {
+            VerifyResponse::Failure {
+                reason: Some(reason),
+            } if reason.contains("expired") || reason.contains("issued later than") => {
+                Err(TokenserverError {
+                    status: "invalid-timestamp",
+                    location: ErrorLocation::Body,
+                    ..Default::default()
+                })
+            }
+            VerifyResponse::Failure { .. } => {
+                Err(TokenserverError::invalid_credentials("Unauthorized"))
+            }
+            VerifyResponse::Okay { issuer, .. } if issuer != self.issuer => {
+                Err(TokenserverError::invalid_credentials("Unauthorized"))
+            }
+            VerifyResponse::Okay {
+                idp_claims: Some(claims),
+                ..
+            } if !claims.token_verified() => {
+                Err(TokenserverError::invalid_credentials("Unauthorized"))
+            }
+            VerifyResponse::Okay {
+                email,
+                idp_claims: Some(claims),
+                ..
+            } => Ok(VerifyOutput {
+                device_id: claims.device_id.clone(),
+                email,
+                generation: claims.generation()?,
+                keys_changed_at: claims.keys_changed_at()?,
+            }),
+            VerifyResponse::Okay { email, .. } => Ok(VerifyOutput {
+                device_id: None,
+                email,
+                generation: None,
+                keys_changed_at: None,
+            }),
+        }
+    }
+}
+
+/// The request sent to the FxA BrowserID verifier for token verification.
+#[derive(Serialize)]
+struct VerifyRequest {
+    assertion: String,
+    audience: String,
+    #[serde(rename(serialize = "trustedIssuers"))]
+    trusted_issuers: [String; 1],
+}
+
+/// The response returned by the FxA BrowserID verifier for a token verification request.
+#[derive(Deserialize, Serialize)]
+#[serde(tag = "status", rename_all = "lowercase")]
+enum VerifyResponse {
+    Okay {
```

Comment:
The old Tokenserver looks specifically for a ["status" field set to "okay"](https://github.com/mozilla-services/tokenserver/blob/master/tokenserver/verifiers.py#L161), but the structure of success and error responses differ, making it possible for serde to distinguish between them without having to parse the "status" field at all.